File: src/main/org/nlogo/agent/TurtleStamp3D.java
package org.nlogo.agent;

import org.nlogo.api.Shape;

public strictfp class TurtleStamp3D
    implements org.nlogo.api.TurtleStamp3D {
  private final String shape;
  private final double xcor;
  private final double ycor;
  private final double zcor;
  private final double size;
  private final double heading;
  private final double pitch;
  private final double roll;
  private final Object color;
  private final double lineThickness;

  TurtleStamp3D(Turtle3D turtle) {
    shape = turtle.shape();
    xcor = turtle.xcor();
    ycor = turtle.ycor();
    zcor = turtle.zcor();
    size = turtle.size();
    heading = turtle.heading();
    pitch = turtle.pitch();
    roll = turtle.roll();
    color = turtle.color();
    lineThickness = turtle.lineThickness();
  }

  TurtleStamp3D(String shape, double xcor, double ycor, double zcor, double size,
                double heading, double pitch, double roll, double color, double lineThickness) {
    this.shape = shape;
    this.xcor = xcor;
    this.ycor = ycor;
    this.zcor = zcor;
    this.size = size;
    this.heading = heading;
    this.pitch = pitch;
    this.roll = roll;
    this.color = Double.valueOf(color);
    this.lineThickness = lineThickness;
  }

  public String shape() {
    return shape;
  }

  public Object color() {
    return color;
  }

  public double heading() {
    return heading;
  }

  public double pitch() {
    return pitch;
  }

  public double roll() {
    return roll;
  }

  public double lineThickness() {
    return lineThickness;
  }

  public double size() {
    return size;
  }

  public double xcor() {
    return xcor;
  }

  public double ycor() {
    return ycor;
  }

  public double zcor() {
    return zcor;
  }

  // stuff we're ignoring for right now
  public boolean hasLabel() {
    return false;
  }

  public String labelString() {
    return "";
  }

  public Object labelColor() {
    return null;
  }

  public boolean hidden() {
    return false;
  }

  public long id() {
    return 0;
  }

  public org.nlogo.api.AgentSet getBreed() {
    throw new UnsupportedOperationException();
  }

  public int getBreedIndex() {
    return 0;
  }

  public World world() {
    return null;
  }

  public org.nlogo.api.Protractor protractor() {
    return null;
  }

  public org.nlogo.api.Patch getPatchHere() {
    return null;
  }

  public void jump(double d) {
  }

  public void heading(double d) {
  }

  public String classDisplayName() {
    return "";
  }

  public double dx() {
    return 0;
  }

  public double dy() {
    return 0;
  }

  public double dz() {
    return 0;
  }

  public Object getVariable(int vn) {
    throw new UnsupportedOperationException();
  }

  public void setVariable(int vn, Object value) {
    throw new UnsupportedOperationException();
  }
}

File: netlogo-core/src/main/agent/TurtleStamp3D.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.agent

import org.nlogo.core.AgentKind
import org.nlogo.api.Color

@annotation.strictfp
case class TurtleStamp3D(shape: String, xcor: Double, ycor: Double, zcor: Double, size: Double,
                         heading: Double, pitch: Double, roll: Double, color: AnyRef, lineThickness: Double)
extends org.nlogo.api.TurtleStamp3D {
  def this(t: Turtle3D) =
    this(t.shape, t.xcor, t.ycor, t.zcor, t.size, t.heading, t.pitch, t.roll, t.color, t.lineThickness)

  override val kind = AgentKind.Turtle
  // stuff we're ignoring for now
  override def hasLabel = false
  override def labelString = ""
  override def labelColor = null
  override def hidden = false
  override def id = 0L
  override def getBreed = unsupported
  override def getBreedIndex = 0
  override def world = null
  override def getPatchHere = null
  override def jump(d: Double) {}
  override def heading(d: Double) {}
  override def classDisplayName = ""
  override def dx = 0
  override def dy = 0
  override def dz = 0
  override def alpha = Color.getColor(color).getAlpha
  override def getVariable(vn: Int) = unsupported
  override def setVariable(vn: Int, value: AnyRef) = unsupported
  override def variables = unsupported

  private def unsupported = throw new UnsupportedOperationException
}
